### PR TITLE
Map GI custom operators

### PR DESCRIPTION
**Pull request by wtritch**
This PR adds mappings from TIMS JSON-to-QTI for GI custom operators.

### CODE REVIEW

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiGetPointExpressionMapper.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    public Element mapExpression(final QtiExpressionMapperSupport mapperSupport, final Document document, final Expression expression) {
+        final CustomGiGetPointExpression customGiGetPointExpression = (CustomGiGetPointExpression) expression;
+
+        final Map<String, String> parameters = new HashMap<>();
+        parameters.put("object", customGiGetPointExpression.getObject());
+        if (isNotBlank(customGiGetPointExpression.getPointIndex())) {
+            parameters.put("point-index", customGiGetPointExpression.getPointIndex());
+        }
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiGetPointExpressionMapper.java]: This is the ignored/optional parameter that we discussed earlier.

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiGetSelectedRegionCountExpressionMapper.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import static org.opentestsystem.ap.item.translation.mappers.qti.scoring.ScoringUtils.customOperator;
+
+/**
+ * This class is responsible for mapping {@link CustomGiGetSelectedRegionCountExpression} instances
+ * to equivalent QTI custom operator elements.
+ */
+@Component
+class CustomGiGetSelectedRegionCountExpressionMapper implements QtiExpressionMapper {
+    // QTI scoring engine uses the specification function name
+    static final String GetSelectedRegionsCountFunctionName = "GETSELECTEDREGIONSCOUNT";
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiGetSelectedRegionCountExpressionMapper.java]: The QTI scoring engine follows the spec for this name.  In the JSON and TDS_ItemScoring, this is named `GETSELECTEDREGIONCOUNT` without the `S`

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiIntersectRegionExpressionMapper.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        final CustomGiIntersectRegionExpression customGiIntersectRegionExpression = (CustomGiIntersectRegionExpression) expression;
+
+        final Map<String, String> parameters = new HashMap<>();
+        parameters.put("object", customGiIntersectRegionExpression.getObject());
+        parameters.put("left-x", customGiIntersectRegionExpression.getLeftX());
+        parameters.put("top-y", customGiIntersectRegionExpression.getTopY());
+        parameters.put("right-x", customGiIntersectRegionExpression.getRightX());
+        parameters.put("bottom-y", customGiIntersectRegionExpression.getBottomY());
+
+        return customOperator(document, customGiIntersectRegionExpression.getCustomType(), customGiIntersectRegionExpression.getCustomFunction(), parameters);
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiIntersectRegionExpressionMapper.java]: This relies on an associated PR in AP_Common to properly name the function: https://github.com/SmarterApp/AP_Common/pull/564

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiIntersectPointExpressionMapper.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    public Element mapExpression(final QtiExpressionMapperSupport mapperSupport, final Document document, final Expression expression) {
+        final CustomGiIntersectPointExpression customGiIntersectPointExpression = (CustomGiIntersectPointExpression) expression;
+
+        final Map<String, String> parameters = new HashMap<>();
+        parameters.put("object", customGiIntersectPointExpression.getObject());
+        parameters.put("x", customGiIntersectPointExpression.getX());
+        parameters.put("y", customGiIntersectPointExpression.getY());
+        parameters.put("tolerance", customGiIntersectPointExpression.getTolerance());
+
+        return customOperator(document, customGiIntersectPointExpression.getCustomType(), customGiIntersectPointExpression.getCustomFunction(), parameters);
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiIntersectPointExpressionMapper.java]: This relies on an associated PR in AP_Common to properly name the function: SmarterApp/AP_Common#564

File: src/test/resources/items/gi-183417.qti.xml [hunk 5]
```diff
@@ -73,7 +73,7 @@
           <qti-base-value base-type="string">
             group
           </qti-base-value>
-          <qti-custom-operator class="qti.sbac.customOperators.UNKNOWN.OPERATOR" definition="operator=customGiGetName"/>
+          <qti-custom-operator class="qti.sbac.customOperators.GRAPHIC.GETNAME" definition="object=@"/>
```
Comment (wtritch) [inline src/test/resources/items/gi-183417.qti.xml]: This demonstrates the updated mappings in our sample mapped QTI XML item.

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiIntersectRegionExpressionMapper.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Override
+    public Element mapExpression(final QtiExpressionMapperSupport mapperSupport, final Document document, final Expression expression) {
+        final CustomGiIntersectRegionExpression customGiIntersectRegionExpression = (CustomGiIntersectRegionExpression) expression;
+
+        final Map<String, String> parameters = new HashMap<>();
+        parameters.put("object", customGiIntersectRegionExpression.getObject());
+        parameters.put("left-x", customGiIntersectRegionExpression.getLeftX());
+        parameters.put("top-y", customGiIntersectRegionExpression.getTopY());
+        parameters.put("right-x", customGiIntersectRegionExpression.getRightX());
```
Comment (esotrope) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiIntersectRegionExpressionMapper.java]: makes me wonder what right Y means

Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/scoring/CustomGiIntersectRegionExpressionMapper.java]: yeah... kind of awkward names for what's basically two corners of a rectangular region:
```
*-----
|        |
|        |
-----*
```

**Pull request merged**